go/vt/hook/hook_test.go

```diff
@@ -38,3 +45,222 @@ func TestExecuteContext(t *testing.T) {
+		{
+			name:     "HOOK_INVALID_NAME",
+			input:    HookResult{ExitStatus: HOOK_INVALID_NAME},
+			expected: "result: HOOK_INVALID_NAME",
+		},
+		{
+			name:     "HOOK_VTROOT_ERROR",
+			input:    HookResult{ExitStatus: HOOK_VTROOT_ERROR},
+			expected: "result: HOOK_VTROOT_ERROR",
+		},
+		{
+			name:     "case default",
+			input:    HookResult{ExitStatus: 42},
+			expected: "result: exit(42)",
+		},
+		{
+			name:     "WithStderr",
+			input:    HookResult{ExitStatus: HOOK_SUCCESS, Stderr: "error"},
+			expected: "result: HOOK_SUCCESS\nstderr:\nerror",
+		},
+		{
+			name:     "WithStderr",
+			input:    HookResult{ExitStatus: HOOK_SUCCESS, Stderr: "error"},
+			expected: "result: HOOK_SUCCESS\nstderr:\nerror",
+		},
+		{
+			name:     "WithStdoutAndStderr",
+			input:    HookResult{ExitStatus: HOOK_SUCCESS, Stdout: "output", Stderr: "error"},
+			expected: "result: HOOK_SUCCESS\nstdout:\noutput\nstderr:\nerror",
+		},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			result := tc.input.String()
+			assert.Equal(t, tc.expected, result)
+		})
+	}
+}
+
+func TestExecuteAsReadPipe(t *testing.T) {
+	vtroot, err := vtenv.VtRoot()
+	require.NoError(t, err)
+
+	cat, err := exec.LookPath("cat")
+	require.NoError(t, err)
+
+	catHookPath := path.Join(vtroot, "vthook", "cat")
+
+	if _, err := os.Lstat(catHookPath); err == nil {
+		require.NoError(t, os.Remove(catHookPath))
+	}
+
+	require.NoError(t, os.Symlink(cat, catHookPath))
+	defer func() {
+		require.NoError(t, os.Remove(catHookPath))
+	}()
+
+	h := NewHook("cat", nil)
+	reader, waitFunc, status, err := h.ExecuteAsReadPipe(strings.NewReader("Hello, World!\n"))
+	require.NoError(t, err)
+	defer reader.(io.Closer).Close()
+
+	output, err := io.ReadAll(reader)
+	require.NoError(t, err)
+	assert.Equal(t, "Hello, World!\n", string(output))
+
+	stderr, waitErr := waitFunc()
+	assert.Empty(t, stderr)
+	assert.NoError(t, waitErr)
+	assert.Equal(t, HOOK_SUCCESS, status)
+}
+
+func TestExecuteAsReadPipeErrorFindingHook(t *testing.T) {
+	h := NewHook("nonexistent-hook", nil)
+	reader, waitFunc, status, err := h.ExecuteAsReadPipe(strings.NewReader("Hello, World!\n"))
+	require.Error(t, err)
+	assert.Nil(t, reader)
+	assert.Nil(t, waitFunc)
+	assert.Equal(t, HOOK_DOES_NOT_EXIST, status)
+}
+
+func TestExecuteAsWritePipe(t *testing.T) {
+	vtroot, err := vtenv.VtRoot()
+	require.NoError(t, err)
+
+	echo, err := exec.LookPath("echo")
+	require.NoError(t, err)
+
+	echoHookPath := path.Join(vtroot, "vthook", "echo")
+
+	if _, err := os.Lstat(echoHookPath); err == nil {
+		require.NoError(t, os.Remove(echoHookPath))
+	}
+
+	require.NoError(t, os.Symlink(echo, echoHookPath))
+	defer func() {
+		require.NoError(t, os.Remove(echoHookPath))
+	}()
+
+	h := NewHook("echo", nil)
+	var writer strings.Builder
+	writerPtr := &writer
+	_, waitFunc, status, err := h.ExecuteAsWritePipe(writerPtr)
+	require.NoError(t, err)
+	defer writerPtr.Reset()
+
+	_, err = writer.Write([]byte("Hello, World!\n"))
```

Comment:
Yeah it seems like these lines of code are introducing the race condition, `writer` is not being locked

Reply:
okay, I'll bring in a patch

Reply:
@frouioui  done